File: AGC/AGC039/A/main.java
import java.util.*;

public class Main {

  public static void main (String[] args) {

    Scanner sc = new Scanner(System.in);

    String S = sc.next();
    long K = sc.nextLong();

    sc.close();

    String[] s = S.split("");
    
    TreeMap<Integer, Integer> map = new TreeMap<>();

    int count = 0;
    map.put(count, 1);
    for (int i = 1; i < s.length; i++) {
      if (s[i - 1].equals(s[i])) {
        int v_ = map.get(count);
        map.put(count, v_ + 1);
      } else {
        count++;
        map.put(count, 1);
      }
    }

    if (map.size() == 1) {
    
      System.out.println(s.length * K / 2);
    
    } else {

      int result = 0;
      int size = map.size();
      
      if (!s[0].equals(s[s.length-1])){
      
        for (int i = 0; i < size; i++) {
          int v_ = map.get(i);
          result += v_ / 2;
        }

        System.out.println(result * K);
      
      } else {

        for (int i = 0; i < size; i++) {
          int v_ = map.get(i);
          result += v_ / 2;
        }
        
        int a1 = map.get(0) / 2 + map.get(size - 1) / 2;
        int a2 = (map.get(0) + map.get(size - 1)) / 2;

        System.out.println(result * K + (a2 - a1) * (K - 1));
        
      }

    }

  }

}

File: AGC/AGC039/A/main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {

  string S;
  cin >> S;

  long K;
  cin >> K;

  vector<int> v;
  v.push_back(1);

  int count = 0;
  
  for (int i = 1; i < S.size(); i++) {
    if (S[i - 1] == S[i]) {
      v[count]++;
    } else {
      count++;
      v.push_back(1);
    }
  }

  if (v.size() == 1) {

    cout << S.size() * K / 2 << endl;

  } else {

    long result = 0;
    int size = v.size();
    
    if (S[0] != S[S.size() - 1]) {
      
      for (int i = 0; i < size; i++) {
        int v_ = v[i];
        result += v_ / 2;
      }

      cout << result * K << endl;

    } else {

      for (int i = 0; i < size; i++) {
        int v_ = v[i];
        result += v_ / 2;
      }
      
      int a1 = v[0] / 2 + v[size - 1] / 2;
      int a2 = (v[0] + v[size - 1]) / 2;

      cout << result * K + (a2 - a1) * (K - 1) << endl;

    }

  }

}